svm/src/transaction_processor.rs

```diff
@@ -1292,6 +1299,165 @@ mod tests {
+
+            // Ensure that the program the instruction data claims to be running is the program that
+            // is actually running.
+            let actual_program_address = *instruction_context
+                .get_last_program_key(&transaction_context)
+                .unwrap();
+            assert_eq!(
+                actual_program_address,
+                PROGRAM_ADDRESSES[current_program_index],
+                "The address of the program at index {} that the instruction data claims to be running ({}) is not the program that is actually running ({})",
+                current_program_index,
+                PROGRAM_ADDRESSES[current_program_index],
+                actual_program_address,
+            );
+
+            if current_program_index == index_of_program_that_must_throw_exception {
+                return Err(InstructionError::Custom(0xdeadbeef));
+            }
+
+            if stack_height < 4 && current_program_index % 2 == 0 {
+                // Every odd program does CPIs unless it has reached the maximum CPI stack depth.
+                let mut last_result = Ok(());
+                for ii in 1..3 {
+                    let next_program_index = current_program_index.checked_add(ii).unwrap();
+                    let next_program_address = PROGRAM_ADDRESSES[next_program_index];
+                    let accounts = (next_program_index..PROGRAM_ADDRESSES.len())
+                        .map(|index| AccountMeta::new_readonly(PROGRAM_ADDRESSES[index], false))
+                        .collect_vec();
+                    last_result = invoke_context.native_invoke(
+                        Instruction::new_with_bytes(
+                            next_program_address,
+                            &[
+                                index_of_program_that_must_throw_exception as u8,
+                                next_program_index as u8,
+                            ],
+                            accounts,
+                        )
+                        .into(),
+                        &[],
+                    );
+                    if last_result.is_err() {
+                        return last_result;
+                    }
+                }
+                return last_result;
+            }
+
+            Ok(())
+        });
+
+        // =======================================================
+        // BEGIN: Create a transaction that calls the mock program
+        // =======================================================
+        let base_program_index = 0;
+        let accounts = (0..PROGRAM_ADDRESSES.len())
+            .map(|index| AccountMeta::new_readonly(PROGRAM_ADDRESSES[index], false))
+            .collect_vec();
+        let message: Message = Message::new(
+            &[Instruction::new_with_bytes(
+                PROGRAM_ADDRESSES[base_program_index],
+                &[
+                    index_of_program_that_should_throw_exception,
```

Comment:
This is how you tell the program ‘I want you to throw an error when you get to the program with this index.’